packages/network/src/logic/proxy/ProxyStreamConnectionServer.ts

```diff
@@ -0,0 +1,109 @@
+
+const logger = new Logger(module)
+
+export interface ProxyStreamConnectionServerOptions {
+    streamPartManager: StreamPartManager
+    nodeToNode: NodeToNode
+    propagation: Propagation
+    node: Node
+    acceptProxyConnections: boolean
+}
+
+interface ProxyConnection {
+    direction: ProxyDirection // Direction is from the clients point of view
+    userId: string
+}
+
+export class ProxyStreamConnectionServer {
+    private readonly streamPartManager: StreamPartManager
+    private readonly nodeToNode: NodeToNode
+    private readonly node: Node
+    private readonly acceptProxyConnections: boolean
+    private readonly connections: Map<StreamPartID, Map<NodeId, ProxyConnection>>
+    private readonly propagation: Propagation
+
+    constructor(opts: ProxyStreamConnectionServerOptions) {
+        this.streamPartManager = opts.streamPartManager
+        this.nodeToNode = opts.nodeToNode
+        this.node = opts.node
+        this.acceptProxyConnections = opts.acceptProxyConnections
+        this.propagation = opts.propagation
+        this.connections = new Map()
+        this.nodeToNode.on(NodeToNodeEvent.PROXY_CONNECTION_REQUEST_RECEIVED, (message,  nodeId) => {
+            this.processHandshakeRequest(message, nodeId)
+        })
+
+        this.nodeToNode.on(NodeToNodeEvent.LEAVE_REQUEST_RECEIVED, (message, nodeId) => {
+            this.processLeaveRequest(message, nodeId)
+        })
+    }
+
+    private async processHandshakeRequest(message: ProxyConnectionRequest, nodeId: NodeId): Promise<void> {
+        const streamPartId = message.getStreamPartID()
+        const isAccepted = this.streamPartManager.isSetUp(streamPartId) && this.acceptProxyConnections
+        if (isAccepted) {
+            if (message.direction === ProxyDirection.PUBLISH) {
+                this.streamPartManager.addInOnlyNeighbor(streamPartId, nodeId)
+                this.addConnection(streamPartId, nodeId, ProxyDirection.PUBLISH, message.userId)
+            } else {
+                this.streamPartManager.addOutOnlyNeighbor(streamPartId, nodeId)
+                this.addConnection(streamPartId, nodeId, ProxyDirection.SUBSCRIBE, message.userId)
+                this.propagation.onNeighborJoined(nodeId, streamPartId)
+            }
+        }
+        await this.nodeToNode.respondToProxyConnectionRequest(nodeId, streamPartId, message.direction, isAccepted)
+    }
+
+    private addConnection(streamPartId: StreamPartID, nodeId: NodeId, direction: ProxyDirection, userId: string): void {
+        if (!this.connections.has(streamPartId)) {
+            this.connections.set(streamPartId, new Map())
+        }
+        this.connections.get(streamPartId)!.set(nodeId, {
+            direction,
+            userId
+        })
+    }
+
+    private processLeaveRequest(message: UnsubscribeRequest, nodeId: NodeId): void {
+        const streamPartId = message.getStreamPartID()
+        if (this.streamPartManager.isSetUp(streamPartId)
```

Comment:
Would it make sense to check whether we have a connection `in this.connections`? Instead of checking the facts from `streamPartManager`.